message_manager.go

```diff
@@ -0,0 +1,324 @@
+	)
+	m.host.Peerstore().RecordLatency(p, time.Since(start))
+	return rpmes, nil
+}
+
+// SendMessage sends out a message
+func (m *messageManager) SendMessage(ctx context.Context, p peer.ID, pmes *pb.Message) error {
+	ctx, _ = tag.New(ctx, metrics.UpsertMessageType(pmes))
+
+	ms, err := m.messageSenderForPeer(ctx, p)
+	if err != nil {
+		stats.Record(ctx,
+			metrics.SentMessages.M(1),
+			metrics.SentMessageErrors.M(1),
+		)
+		logger.Debugw("message failed to open message sender", "error", err, "to", p)
+		return err
+	}
+
+	if err := ms.SendMessage(ctx, pmes); err != nil {
+		stats.Record(ctx,
+			metrics.SentMessages.M(1),
+			metrics.SentMessageErrors.M(1),
+		)
+		logger.Debugw("message failed", "error", err, "to", p)
+		return err
+	}
+
+	stats.Record(ctx,
+		metrics.SentMessages.M(1),
+		metrics.SentBytes.M(int64(pmes.Size())),
+	)
+	return nil
+}
+
+func (m *messageManager) messageSenderForPeer(ctx context.Context, p peer.ID) (*messageSender, error) {
+	m.smlk.Lock()
+	ms, ok := m.strmap[p]
+	if ok {
+		m.smlk.Unlock()
+		return ms, nil
+	}
+	ms = &messageSender{p: p, m: m, lk: newCtxMutex()}
+	m.strmap[p] = ms
+	m.smlk.Unlock()
+
+	if err := ms.prepOrInvalidate(ctx); err != nil {
+		m.smlk.Lock()
+		defer m.smlk.Unlock()
+
+		if msCur, ok := m.strmap[p]; ok {
+			// Changed. Use the new one, old one is invalid and
+			// not in the map so we can just throw it away.
+			if ms != msCur {
+				return msCur, nil
+			}
+			// Not changed, remove the now invalid stream from the
+			// map.
+			delete(m.strmap, p)
+		}
+		// Invalid but not in map. Must have been removed by a disconnect.
+		return nil, err
+	}
+	// All ready to go.
+	return ms, nil
+}
+
+type messageSender struct {
+	s  network.Stream
+	r  msgio.ReadCloser
+	lk ctxMutex
+	p  peer.ID
+	m  *messageManager
+
+	invalid   bool
+	singleMes int
+}
+
+// invalidate is called before this messageSender is removed from the strmap.
+// It prevents the messageSender from being reused/reinitialized and then
+// forgotten (leaving the stream open).
+func (ms *messageSender) invalidate() {
+	ms.invalid = true
+	if ms.s != nil {
+		_ = ms.s.Reset()
+		ms.s = nil
+	}
+}
+
+func (ms *messageSender) prepOrInvalidate(ctx context.Context) error {
+	if err := ms.lk.Lock(ctx); err != nil {
+		return err
+	}
+	defer ms.lk.Unlock()
+
+	if err := ms.prep(ctx); err != nil {
+		ms.invalidate()
+		return err
+	}
+	return nil
+}
+
+func (ms *messageSender) prep(ctx context.Context) error {
+	if ms.invalid {
+		return fmt.Errorf("message sender has been invalidated")
+	}
+	if ms.s != nil {
+		return nil
+	}
+
+	// We only want to speak to peers using our primary protocols. We do not want to query any peer that only speaks
+	// one of the secondary "server" protocols that we happen to support (e.g. older nodes that we can respond to for
+	// backwards compatibility reasons).
+	nstr, err := ms.m.host.NewStream(ctx, ms.p, ms.m.protocols...)
```

Comment:
this is the only place that the message sender uses it's reference to the manager, and also the only place the host is really made use of.

Can we provide messageSender's with a closure for making new streams, rather than a full reference to the manager? making it clear that responsibility is `manger->sender(s)` will be clearer code than having back references.

Reply:
No problem here. Another copy-paste, this was previously a reference to the entire `*IpfsDHT`.

Reply:
We can, but we also need the host for the latency measurements above. I'm not sure who is using those metrics or why (perhaps @Stebalien has some ideas), so I left things the same (previously there was a reference to the `IpfsDHT` here).